src/utilities/export-pdf-helper.ts

```diff
@@ -0,0 +1,242 @@
+            ],
+        });
+
+        if (_.isNil(mainData)) {
+            throw new NotFoundError(`Could not find topic ${topic.id} to export.`);
+        }
+
+        const studentGradeLookup: {
+            [userId: number]: {
+                number: number;
+                srcdoc: string;
+                attachments: {url: string; name: string}[] | undefined;
+                effectiveScore: number;
+                legalScore: number;
+            }[];
+        } = {};
+
+        await mainData.questions?.sequentialAsyncForEach(async (question) =>
+            await question.grades?.sequentialAsyncForEach(async (grade) => {
+                if (_.isNil(studentGradeLookup[grade.userId])) {
+                    studentGradeLookup[grade.userId] = [];
+                }
+
+                const influencingWorkbook = grade.lastInfluencingCreditedAttemptId ?? grade.lastInfluencingAttemptId;
+
+                if (_.isNil(influencingWorkbook)) {
+                    // If we can't find a workbook, we can't display anything for the student attempt.
+                    studentGradeLookup[grade.userId].push({
+                        number: question.problemNumber,
+                        attachments: [],
+                        srcdoc: 'There is no record of this student attempting this problem.',
+                        effectiveScore: grade.effectiveScore,
+                        legalScore: grade.legalScore,
+                    });
+
+                    return;
+                }
+
+                const gradeInstanceAttachments = await StudentWorkbook.findOne({
+                    where: {
+                        id: influencingWorkbook,
+                        active: true,
+                    },
+                    attributes: ['id', 'submitted', 'randomSeed'],
+                    include: [
+                        {
+                            model: StudentGradeInstance,
+                            as: 'studentGradeInstance',
+                            attributes: ['id', 'webworkQuestionPath', 'randomSeed'],
+                            required: false,
+                            where: {
+                                active: true,
+                            },
+                            include: [
+                                {
+                                    model: ProblemAttachment,
+                                    as: 'problemAttachments',
+                                    attributes: ['id', 'cloudFilename', 'userLocalFilename', 'updatedAt'],
+                                    where: {
+                                        active: true,
+                                    }
+                                }
+                            ]
+                        },
+                    ]
+                });
+
+                const obj = {
+                    sourceFilePath: gradeInstanceAttachments?.studentGradeInstance?.webworkQuestionPath ?? question.webworkQuestionPath,
```

Comment:
With the problem path having been added to the workbook - does it make sense here to rely on the workbook as the source of truth? Then falling back to grade instance, and finally to the topic question as a last resort?